### PR TITLE
Made creature/hero dots on panel minimap larger

### PR DESCRIPTION
**Pull request by spartahawk**
Someone in the forum asked if creature/hero dots can be made larger, I thought it was a good idea. 
Sometimes the single pixel isn't so easily visible.

Dots were previously a single pixel and difficult to see in some instances in the mini-map. Now they are 2x2 for regular creatures and 3x3 for highlighted creatures (under possession I believe -- it's a cross like a plus)

### CODE REVIEW

**Comment by Loobinex** [comment]
I never experienced as problematic myself, but I'll have a look. Tomorrow, it's 01:41 AM now over here.

**Comment by Loobinex** [comment]
I've looked into it and I see what the user means in [this topic](https://keeperklan.com/threads/6889-Minimap-Unit-dots-only-1-pixel-big-(almost-microscopic!)).
I looked at it in different ingame resolutions and zoom-levels and I think the main thing is the scaling of the pixel across zoom levels. The correct size of the pixel - like in the original game - is that is is small enough that you can see 9 separate units per tile (so one per subtile). If they become larger than that it will start to look messy when you've got multiple units as they can no longer pass through the center of a corridor and they will clip through walls. You also can no longer see how many imps are mining gems.
See this screenshot from your prototype:
mapzoom
Units are here half a corridor wide, which is too much.

I also dislike that from your prototype the size of the units on the main map is different from the minimap on the same zoom-level.

To improve on the issue I think the size could be different on maximum-zoom on the minimap, where they actually are too small. Improve them there to be the same size relative to the second biggest zoom level, but always being able to see 9 units per tile.

**Comment by spartahawk** [comment]
Good points. I'll see if I can scale it to meet those requirements.

For the panel mini-map, zooming out past a certain point they'll need to remain 1 pixel to be distinct enough for 9 creatures visible per tile. When you're zoomed out that much though you might as well use the Map Screen though if you're looking for creatures, so 1 pixel at those zooms is probably best. Personally I'd always rather more easily see something (2x2) than have the granularity (1-pixel at the furthest zooms).  But maybe that's a good reason to have a user choice. It would likely be a command-line option since I'm not sure if I can change the in-game display options panel.

The original game (at least my GOG version) uses 2x2 dots at every zoom, which is what I implemented. This causes the issues you mention at certain zooms. But we have options so why not improve it.

In the original (GOG) game in the Map Screen, on my display, if three imps (each 2x2) are digging at one tile, it looks like a solid 2x6 line in the Map Screen, at least on my display. So they don't overlap, but they're not distinct. I'll make them 2x2, let me know how that looks on your display once I've done it (I'll comment again). I finally found the code where the Map Screen drawing is implemented so I'll try tweaking it.

Clipping walls etc is inevitable when you're using 2x2 pixels based around a single pixel, unless you're dynamically changing how they center around the actual position pixel. I have it so the additional 3 pixels are to the right and below, and not symmetrically surrounding the center. I suppose I could change it based on where exactly the creature is but for now I'd just leave the 1-pixel if granularity and actual position (at the expense of visibility) is most important to the user.

**Comment by spartahawk** [comment]
Do you know how where in the code the keeperfx.cfg values are read?

Minimap zoom-3 (the most useful zoom I think) is actually slightly smaller
in scale than the full map screen, so while a 2x2 dot is problematic in
zoom-3, it causes overlap in zoom-3. So, although they're not the same
scale, to make them the most consistent, zoom-3 would have 2x2
(overlapping) pixels.

We have 2 options for the creature dots at zoom level 3, if we want
distinct creature "dots", none of which will look consistent with the full
map screen. Either single pixels, which is clearly distinct but not very
visible (and wouldn't match the full map screen'ss 2x2), or a 2x1 vertical,
horizontal, or diagonal line, which works but if your rotation isn't
diagonal to it, you lose distinction in one direction.

But the minimap zoom-3 is probably not so important anyway if you can go to
the map screen often.

What do you prefer for creatures at zoom-3: single pixels, 2x1 lines, or
the 2x2 dots you have screenshots of above?



​

**Comment by spartahawk** [comment]
If you want to compare the two without re-compiling yourself, here they are.

[zoom-3 1-pixel](https://drive.google.com/open?id=1t52_Zcfsqw0jgKmaNvtDryTJKwgxHivG)

[zoom-3 2x1 vertical line](https://drive.google.com/open?id=1TLs5wWlHnk-eCFBHg2Ul3SOQEAjPGMSM)

**Comment by Loobinex** [comment]
Keep in mind that what level of zoom is most useful also depends on the screen size and resolution of the player.
Personally I use lvl2 the most by the way, as on three it becomes to tiny for me. And indeed it matters a lot if you play north as up or at an angle. I think after a while most people will end up looking north as as soon as you open a map this is where you'll end up, but it should look nice on both orientations.

I really dislike the look of lines, I think the units on the map should be square (or round), not rectangular.
And perhaps we can abandon the distinctiveness of the units on zoom level 3 and beyond. What I really do dislike though is when they clip through walls .

**Comment by Loobinex** [comment]
I think I quite like that zoom-3 1 pixel version. It's different than I'm used to, which is always scary of course, but this could be a winner.
Do you like it as well?

If you're in no rush to commit, let me hang on to this version for a bit to see and use it in an actual game.

**Comment by spartahawk** [comment]
I agree that symmetrical (not lines) is the way to go.

KeeperFX has been 1-pixel at every zoom on the panel minimap up until these changes if I'm not mistaken. If you play a build of the code on the master branch now that's what you get.

**Comment by spartahawk** [comment]
Apparently I accidentally closed the pull request. I can submit a new one
some time.

Thanks for the thoughts.

**Comment by spartahawk** [comment]
I should really get a bigger, high-res screen so I can see how the game looks with that setup. Maybe you actually haven't seen the 1-pixel at zoom 3. I assumed you had since KeeperFX has been that way for me (all zoom levels being 1 pixel)

**Comment by Loobinex** [comment]
I am still rocking my screen that was very high end a long time ago, with a 1280x1024 resolution. And indeed in that resolution on zoom-level 3 I see no difference between your prototype and the original. Zoom levels 1 and 2 do look different between the two versions, with substantially larger but still fully functional units in your prototype.
On 640x480 resolution however (the resolution GOG also plays at) there is some clipping now with the prototype.

So perhaps this issue is caused only by the way the sidebar is scaling to higher resolutions, perhaps it is forgetting to scale the units on the (mini)-map.

**Comment by spartahawk** [comment]
I did the scaling up too, only so that the tiny dots looked more
appropriately sized compared to rooms. I just upped them to 5x5 and 3x3
respectively. Frankly I think 2x2 for levels 1 and 2 would work fine, and
that would reduce clipping.

See if you like the map screen dots size better too.

On Jan 22, 2018 4:42 PM, "Loobinex" <notifications@github.com> wrote:

I am still rocking my screen that was very high end a long time ago, with a
1280x1024 resolution. And indeed in that resolution on zoom-level 3 I see
no difference between your prototype and the original. Zoom levels 1 and 2
do look different between the two versions, with substantially larger but
still fully functional units in your prototype.
On 640x480 resolution however (the resolution GOG also plays at) there is
some clipping now with the prototype.

So perhaps this issue is caused only by the way the sidebar is scaling to
higher resolutions, perhaps it is forgetting to scale the units on the
(mini)-map.

—
You are receiving this because you modified the open/close state.

Reply to this email directly, view it on GitHub
<https://github.com/Loobinex/keeperfx-unofficial/pull/5#issuecomment-359612751>,
or mute the thread

**Comment by spartahawk** [comment]
Well to clarify, I did nothing that was based on the screen resolution.
That's the issue I'm sure. Not sure I want to factor that in either, since
I can't test it myself.

**Comment by Loobinex** [comment]
Sure you can test that. I don't know what your screen resolution is, but I'm sure it's larger than 640x480 right?
Configure your keeperfx config like this:
INGAME_RES=1280x1024x32 640x480x32

With the first one whatever your native resolution is. Ingame, simply press Alt+R to see the game in a different resolution.

Right now, on each resolution the menubar stays the same size, but the size of the units on the map changes size depending on the screen resolution. That's not right,...

**Comment by spartahawk** [comment]
I think if someone's playing in 640x480, they either don't care how the game looks anyway, or they can't afford to care.

I'll look into scaling based on resolution though.

**Comment by spartahawk** [comment]
On second thought, maybe some people enjoy a more pixelated experience. Especially if they want the 1995 feel they remember.

**Comment by Loobinex** [comment]
Actually the difference isn't that big between 640x480 and higher resolutions, as all the sprites are based on that resolution. People with older pc's which have more trouble running KeeperFX can easily decide to keep it at that resolution to have it run more smoothly. Than there are some people who might like the original look, but I think the largest group are the people who leave the default unchanged because they don't know how to change it.

File: src/frontmenu_ingame_map.c [hunk 1]
```diff
@@ -356,53 +356,52 @@ void pannel_map_draw_creature_dot(long mapos_x, long mapos_y, RealScreenCoord ba
 {
     // actual position single pixel
     pannel_map_draw_pixel(mapos_x+basepos, mapos_y+basepos, col);
-    // altered to be only for LESS than 512, meaning only for zooms 1 or 2
-    // to apply to zoom 512 as well, change to <=, although this causes
+    // cab be altered to be only for LESS than 512, meaning only for zooms 1 or 2.
```
Comment (Loobinex) [inline src/frontmenu_ingame_map.c]: * can

**Comment by Loobinex** [comment]
zl5

The higher zoomlevels are now substantially worse than they used to be.
I'm starting to feel bad about giving so much negative feedback here, but it seems to be an issue with a lot more complexity than initially assumed.

**Comment by spartahawk** [comment]
Woah, is all the beige color on the new one from creatures? Or are the
creatures not showing up in that screenshot? It's not even white, it's
beige. Don't know why it would be a different color.

I'm assuming that's on 640x480.

Either way, zooms 4 (pictured) and 5 shouldn't be any different than our
master version. It should be doing the exact same thing. If it isn't it's
just a mistake.

Negative feedback welcome :)

It really shouldn't be hard to fix for resolutions after studying a bit
more.



On Jan 22, 2018 6:11 PM, "Loobinex" <notifications@github.com> wrote:

[image: zl5]
<https://user-images.githubusercontent.com/13840686/35252777-47c1126e-ffe2-11e7-88b6-f3816bd25db5.png>

The higher zoomlevels are now substantially worse than they used to be.
I'm starting to feel bad about giving so much negative feedback here, but
it seems to be an issue with a lot more complexity than initially assumed.

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub
<https://github.com/Loobinex/keeperfx-unofficial/pull/5#issuecomment-359634683>,
or mute the thread

**Comment by Loobinex** [comment]
Both screenshots are from the same saved game, one with the original game the other with the build from your branch. Unless specified I always send my screenshots from 1280x1024 format, and I did here.
And yes those are hero units.

**Comment by spartahawk** [comment]
Do you mind sharing that saved game with me so I can test it out? Thanks!

**Comment by spartahawk** [comment]
Either I'm just unable to reproduce the issue or you're using an older commit. First two are 1280x1024 and the second two are my own screens's native resolution, 1360x768.

1280x1024zoom4
1280x1024zoom5
1360x768zoom4
1360x768zoom5

**Comment by spartahawk** [comment]
Here are the up-to-date files you can put in your keeperfx directory if it saves you time.
https://drive.google.com/open?id=1Rdz4kjuj6tjo1YLVDzWrZaK-bwzunCVV

**Comment by Loobinex** [comment]
Yeah, my mistake. I did not sync before I made the latest build. Looking at your latest prototype with the resolution factored in it all seems all right.
Are there any downsides to this implementation?

I'll look at it more thoroughly later tonight.

**Comment by spartahawk** [comment]
Only downsides:

The full map screen will always use 2 pixels, and this makes the dots too big for at least your liking in the low res version. But I think it's a best-case scenario until I can factor in resolution. 1-pixel is too tiny for ease of use I think.

Zoom-3 uses 1-pixel dots, which at my resolution makes them not very visible. But I'll leave it that way so that there's no overlap or clipping, again until I can factor in resolution (or provide options to the user).

Considering the options right now, those compromises are acceptable enough to merge. And I can merge future improvements soon. Or look at other higher-priority stuff now.

**Comment by Loobinex** [comment]
Sounds to me like you prefer the 2pixel dots on zoom3, so do I. Sure I'd prefer units to be distinguishable on the map, but everything is a trade of. I also like the units on zoom 1 to be a bit bigger.
What do you think about this?

**Comment by spartahawk** [comment]
Perfect. That can be done.

**Pull request merged**